src/core/memory/freezer.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "common/assert.h"
+#include "core/core.h"
+#include "core/core_timing_util.h"
+#include "core/memory.h"
+#include "core/memory/freezer.h"
+
+namespace Memory {
+
+constexpr s64 MEMORY_FREEZER_TICKS = static_cast<s64>(Core::Timing::BASE_CLOCK_RATE / 60);
+
+namespace {
+
+u64 MemoryReadWidth(u8 width, VAddr addr) {
+    switch (width) {
+    case 1:
+        return Read8(addr);
+    case 2:
+        return Read16(addr);
+    case 4:
+        return Read32(addr);
+    case 8:
+        return Read64(addr);
+    default:
+        UNREACHABLE();
+        return 0;
+    }
+}
+
+void MemoryWriteWidth(u8 width, VAddr addr, u64 value) {
+    switch (width) {
+    case 1:
+        Write8(addr, static_cast<u8>(value));
+        break;
+    case 2:
+        Write16(addr, static_cast<u16>(value));
+        break;
+    case 4:
+        Write32(addr, static_cast<u32>(value));
+        break;
+    case 8:
+        Write64(addr, value);
+        break;
+    default:
+        UNREACHABLE();
+    }
+}
+
+} // Anonymous namespace
+
+Freezer::Freezer(Core::Timing::CoreTiming& core_timing) : core_timing(core_timing) {
+    event = core_timing.RegisterEvent(
+        "MemoryFreezer::FrameCallback",
+        [this](u64 userdata, s64 cycles_late) { FrameCallback(userdata, cycles_late); });
+    core_timing.ScheduleEvent(MEMORY_FREEZER_TICKS, event);
+}
+
+Freezer::~Freezer() {
+    core_timing.UnscheduleEvent(event, 0);
+}
+
+void Freezer::SetActive(bool active) {
+    if (!this->active.exchange(active)) {
+        FillEntryReads();
+        core_timing.ScheduleEvent(MEMORY_FREEZER_TICKS, event);
+        LOG_DEBUG(Common_Memory, "Memory freezer activated!");
+    } else {
+        LOG_DEBUG(Common_Memory, "Memory freezer deactivated!");
+    }
+}
+
+bool Freezer::IsActive() const {
+    return active.load();
```

Comment:
You likely just want:

```suggestion
    return active.load(std::memory_order_relaxed);
```

rather than enforcing the strictest form of atomic ordering.